The XSLT stylesheet file is now initialized to an env.File() Node, such that dependencies work correctly in hierarchical builds.

hwmaier opened the pull request:
Re-applying a patch I submitted already in 2013 and was accepted back then. Refer to commit e63d742d2026 in the old Mercurial https://bitbucket.org/dirkbaechle/scons_docbook repo.

Somehow during upgrade and integration of the docbook this must have got lost.

I only stumbled across this because I finally upgraded from Scons 3 to 4 after all those years.

## Contributor Checklist:

* [x] I have created a new test or updated the unit tests to cover the new/changed functionality.
* [x] I have updated `CHANGES.txt` (and read the `README.rst`)
* [x] I have updated the appropriate documentation

mwichmann commented:
This sounds reasonable enough, though the bitbucket repo pointed to here is gone in their purge. @dirkbaechle ?

bdbaddog commented:
@hwmaier You removed the developer checklist which is in the pull request template. 
I've restored it . 
Please fill it out.

bdbaddog commented:
What breaks without this change?

hwmaier commented:
Given that this change already was committed and accepted earlier, I felt it would not be necessary for those formalities.

This is the log of the original branch (which I still have a local copy of) on bitbucket:


changeset:   11:e63d742d2026
parent:      9:b0e290b5d5ef
user:        dirkbaechle
date:        Fri Oct 04 11:13:12 2013 +0200
summary:     - a patch, contributed by Henrik Maier: the XSLT stylesheet file is now initialized to an env.File() Node, such that

changeset:   9:b0e290b5d5ef
user:        dirkbaechle
date:        Mon Apr 22 23:47:53 2013 +0200
summary:     - fixed the lxml support for xincludes and xsl transformations

changeset:   8:1686cc7cdee5
user:        dirkbaechle
date:        Sun Mar 24 19:43:09 2013 +0100
summary:     - always resolve entities when reading files with libxml2

changeset:   7:0afc056ca355
user:        dirkbaechle
date:        Sat Mar 09 12:27:04 2013 +0100
summary:     - small fix: If the input file for DocbookMan doesn't exist yet (it is generated during the build), we can't derive the target name from its contents. So we have to rely on the given target name instead.

changeset:   6:0ab3bdbaab05
user:        dirkbaechle
date:        Sat Mar 09 11:53:17 2013 +0100
summary:     - changed the interface of DocbookXInclude, it now always requires a target AND a source parameter

changeset:   5:3ff5ccf6e4fc
user:        dirkbaechle
date:        Wed Nov 14 23:50:09 2012 +0100
summary:     - fix by Pawel Tomulik: changed terminator symbol for XML dependencies detection

changeset:   4:74f0af5781ec
user:        dirk
date:        Tue Jul 05 01:06:17 2011 +0200
summary:     - support for preferring xsltproc over libxml2 and lxml (environment variable DOCBOOK_PREFER_XSLTPROC)

changeset:   3:f28ed172f9aa
user:        dirk
date:        Tue Jul 05 00:52:21 2011 +0200
summary:     - added support for implicit dependencies, based on system entities in XML files

changeset:   2:41057821ac4d
user:        dirk
date:        Wed Jun 08 20:56:17 2011 +0200
summary:     - improved dependency scanning

changeset:   1:e7212442c66a
user:        Dirk Baechle <dl9obn@darc.de>
date:        Tue Jan 04 00:53:35 2011 +0100
summary:     - added dependencies for the used XSL file to all builders

changeset:   0:e8bbba1f699e
user:        Dirk Baechle <dl9obn@darc.de>
date:        Thu Dec 30 15:43:28 2010 +0100
summary:     - first version

bdbaddog commented:
You'd be incorrect.
Clearly undoing the change didn't break any tests.. 
So that's why we like to add tests with code changes.

All PR's go through the same process, excluding some housekeeping by the maintainer..

hwmaier commented:
> What breaks without this change?

Sorry, forgot to explain this.

If you have a main `SConstruct` file in the top level directory with this content:

```
SConscript('doc/SConscript', 'env')
```

and then a SConscript file in the `doc` subdir with:

```
env.DocbookXslt('xml/index.aml', 'xml/indexpage.xml', xsl = 'doxy2aml.xsl')
```
and the XSL stylesheet `doxy2aml.xsl` resides in `doc\doxy2aml.xsl` then the style sheet will not be found because the `doc` subdirectory is not applied to the stylesheet's file node in this hierarchical build scenario. Once the patch is applied, then the subdir is applied to the node and the style sheet is passed to the xslt processor with correct path name.

hwmaier commented:
So I added a test case which I think is good anyway as it also explains better what goes wrong.

Without the patch:
```
C:\temp\scons>python runtest.py  test\Docbook\basic\xsltsubdir\xsltsubdir.py
1/1 (100.00%) C:\\Python39\\python.exe test\Docbook\basic\xsltsubdir\xsltsubdir.py
File doesn't exist: `C:\Users\henrik\AppData\Local\Temp\testcmd.8360.67of99er\subdir/out.xml'
FAILED test of C:\temp\scons\scripts\scons.py
        at line 578 of C:\temp\scons\testing\framework\TestCommon.py (must_not_be_empty)
        from line 42 of C:\temp\scons\test\Docbook\basic\xsltsubdir\xsltsubdir.py
```

with the applied patch:
```
C:\temp\scons>python runtest.py  test\Docbook\basic\xsltsubdir\xsltsubdir.py
1/1 (100.00%) C:\\Python39\\python.exe test\Docbook\basic\xsltsubdir\xsltsubdir.py
PASSED
```

I also did notice that the docbook tool (`__build_lxml` function, `etree.parse` call in this case) does not write exceptions on the console if something goes wrong internally. Not sure if that is intended, but it makes fault finding harder as the Scons will not show an error on the console and the DocbookXslt tool silently fails with simply not generating any output. If I wrap `etree.parse` in a try/except block and then print the exception, we see what is going wrong.

bdbaddog commented:
@hwmaier - Hiding the etree.parse error was most likley not intentional. If you can capture and output that, it'd be a great addition.

bdbaddog commented on test/Docbook/basic/xsltsubdir/xsltsubdir.py:
```diff
@@ -0,0 +1,57 @@
+test = TestSCons.TestSCons()
+
+try:
+    import lxml
```
Please add this to get sider to ignore unused lxml.

`# noqa: F401`

hwmaier replied:
@bdbaddog Done. I was not aware of this and copied most test logic from the xlst.py test

hwmaier commented:
> @hwmaier - Hiding the etree.parse error was most likley not intentional. If you can capture and output that, it'd be a great addition.
This may be more of an issue with lxml than Scons.Docbook.  I don't feel comfortable to dig deeper into this as I never used lxml and don't know the correct API usage and method or dealing with potential errors.

bdbaddog commented on test/Docbook/basic/xsltsubdir/image/subdir/SConscript:
```diff
@@ -0,0 +1,8 @@
+Import('env')
+env = env.Clone(tools=['docbook'])
```
Curious. Any reason not to just initialize docbook in the top environment and skip the clone()?

hwmaier replied:
I copied the test case init from our build scripts which initialise a complex environment at the top level and then have subscripts with more specialised environments which shall be kept segregated from each other.

It was in this hierarchical configuration that I encountered the issue and thats what I based the test case on.

bdbaddog commented:
Merged.
Note I removed the clone from the SConscript and just set `tools=['docbook']` in the Environment in SConstruct and test was still effective.

bdbaddog commented:
Another minor note. For next time, we ask that people add their names to CHANGES in alphabetical by last name order. Just to keep it organized.

Thanks for the fix!
